### PR TITLE
Adjust RequestQueue so that it can be used for onion client

### PR DESCRIPTION
**Pull request by kurnevsky**

### CODE REVIEW

File: src/toxcore/dht/request_queue.rs [hunk 1]
```diff
@@ -1,71 +1,85 @@
...
+            return None;
         }
 
-        match self.ping_map.remove(&(pk, ping_id)) {
-            Some(time) if clock_elapsed(time) <= self.timeout => true,
-            _ => false,
+        if let Entry::Occupied(entry) = self.ping_map.entry(ping_id) {
+            let &(time, ref data) = entry.get();
+            if clock_elapsed(time) <= self.timeout && cond(data) {
+                Some((entry.remove_entry().1).1)
```
Comment (kpp) [inline src/toxcore/dht/request_queue.rs]: Rewrite `(entry.remove_entry().1).1` to temp variables

File: src/toxcore/dht/request_queue.rs [hunk 2]
```diff
@@ -1,71 +1,85 @@
...
         if ping_id == 0 {
-            return false
+            return None;
         }
 
-        match self.ping_map.remove(&(pk, ping_id)) {
-            Some(time) if clock_elapsed(time) <= self.timeout => true,
-            _ => false,
+        if let Entry::Occupied(entry) = self.ping_map.entry(ping_id) {
+            let &(time, ref data) = entry.get();
```
Comment (kpp) [inline src/toxcore/dht/request_queue.rs]: Can you try to write it without `let &()`?

File: src/toxcore/dht/request_queue.rs [hunk 3]
```diff
@@ -1,71 +1,85 @@
...
+        } else {
+            None
         }
     }
 
     /// Remove timed out request IDs.
     pub fn clear_timed_out(&mut self) {
         let timeout = self.timeout;
-        self.ping_map.retain(|&_, &mut time|
+        self.ping_map.retain(|&_, &mut (time, _)|
```
Comment (kpp) [inline src/toxcore/dht/request_queue.rs]: `&mut`?

File: src/toxcore/dht/request_queue.rs [hunk 4]
```diff
@@ -79,80 +93,69 @@ mod tests {
...
-        let ping_id = queue.new_ping_id(pk);
+        let ping_id = queue.new_ping_id(7);
 
-        assert!(queue.ping_map.contains_key(&(pk, ping_id)));
+        assert_eq!(queue.ping_map[&ping_id].1, 7);
     }
 
     #[test]
     fn check_ping_id() {
-        crypto_init().unwrap();
```
Comment (kpp) [inline src/toxcore/dht/request_queue.rs]: gen_ping_id calls to random_u32 which requires crypto_init()?

File: src/toxcore/dht/request_queue.rs (hunk 2 above)
Comment (kurnevsky) [reply src/toxcore/dht/request_queue.rs]: Why? And how?

File: src/toxcore/dht/request_queue.rs (hunk 3 above)
Comment (kurnevsky) [reply src/toxcore/dht/request_queue.rs]: Yes, it's the signature of `retain` method - it takes such closure.

File: src/toxcore/dht/request_queue.rs (hunk 4 above)
Comment (kurnevsky) [reply src/toxcore/dht/request_queue.rs]: You are right, fixed.

File: src/toxcore/dht/request_queue.rs (hunk 1 above)
Comment (kurnevsky) [reply src/toxcore/dht/request_queue.rs]: done

File: src/toxcore/dht/request_queue.rs (hunk 2 above)
Comment (kpp) [reply src/toxcore/dht/request_queue.rs]:
```
            let (time, data) = entry.get();
            if clock_elapsed(*time) <= self.timeout && cond(data) {
```

File: src/toxcore/dht/request_queue.rs [hunk 5]
```diff
@@ -1,71 +1,86 @@
...
             clock_elapsed(time) <= timeout
         );
     }
+
+    /// Get not timed out requests stored in this `RequestQueue`.
+    pub fn get_values(&self) -> impl Iterator<Item = (Instant, &T)> {
+        self.ping_map
+            .values()
+            .filter(move |(time, _)| clock_elapsed(*time) <= self.timeout)
+            .map(|&(time, ref data)| (time, data))
```
Comment (kpp) [inline src/toxcore/dht/request_queue.rs]: .map(|(time, data)| (*time, data))

Comment (kurnevsky) [reply src/toxcore/dht/request_queue.rs]: Why?

File: src/toxcore/dht/request_queue.rs (hunk 2 above)
Comment (kurnevsky) [reply src/toxcore/dht/request_queue.rs]: Why?

Comment (kpp) [reply src/toxcore/dht/request_queue.rs]: Because I think it is easier to read than your version.

File: src/toxcore/dht/request_queue.rs (hunk 5 above)
Comment (kpp) [reply src/toxcore/dht/request_queue.rs]: Because I think it is easier to read than your version.

**Pull request merged**